File: frontend/src/backend/planService.js
import { fetchConfig, appFetch } from './appFetch';

export const getDailyPlan = (date, onSuccess, onErrors) => {
    appFetch(`/plans/daily?date=${date}`, fetchConfig('GET'), onSuccess, onErrors);
};

export const getWeeklyPlan = (startDate, onSuccess, onErrors) => {
    appFetch(`/plans/weekly?startDate=${startDate}`, fetchConfig('GET'), onSuccess, onErrors);
};

export const createTrainingSession = (trainingSessionData, onSuccess, onErrors) => {
    appFetch(`/plans/create-training-session`, fetchConfig('POST', trainingSessionData), onSuccess, onErrors);
};

export const createNutritionPlan = (nutritionPlanData, onSuccess, onErrors) => {
    appFetch(`/plans/create-nutrition-plan`, fetchConfig('POST', nutritionPlanData), onSuccess, onErrors);
};

export const createRestPlan = (restPlanData, onSuccess, onErrors) => {
    appFetch(`/plans/create-rest-plan`, fetchConfig('POST', restPlanData), onSuccess, onErrors);
};

export const updateTrainingBlockDone = (trainingBlockData, onSuccess, onErrors) => {
    appFetch(`/plans/update-training-block-done`, fetchConfig('POST', trainingBlockData), onSuccess, onErrors);
};

export const updateNutritionPlanDone = (nutritionPlanData, onSuccess, onErrors) => {
    appFetch(`/plans/update-nutrition-plan-done`, fetchConfig('POST', nutritionPlanData), onSuccess, onErrors);
};

export const updateRestPlanDone = (restPlanData, onSuccess, onErrors) => {
    appFetch(`/plans/update-rest-plan-done`, fetchConfig('POST', restPlanData), onSuccess, onErrors);
};

export const rescheduleTrainingSession = (rescheduleData, onSuccess, onErrors) => {
    appFetch(`/plans/reschedule-training-session`, fetchConfig('POST', rescheduleData), onSuccess, onErrors);
};

export const getAthleteDailyPlan = (athleteId, date, onSuccess, onErrors) => {
    appFetch(`/plans/athletes/${athleteId}/daily?date=${date}`, fetchConfig('GET'), onSuccess, onErrors);
};

export const getAthleteWeeklyPlan = (athleteId, startDate, onSuccess, onErrors) => {
    appFetch(`/plans/athletes/${athleteId}/weekly?startDate=${startDate}`, fetchConfig('GET'), onSuccess, onErrors);
};

export const getNotifications = (onSuccess, onErrors) => {
    appFetch(`/plans/notifications`, fetchConfig('GET'), onSuccess, onErrors);
};

export const markNotificationAsRead = (id, onSuccess, onErrors) => {
    appFetch(`/plans/notifications/${id}/read`, fetchConfig('POST'), onSuccess, onErrors);
};

export const acceptReadjustment = (notificationId, params, onSuccess, onErrors) => {
    appFetch(`/plans/notifications/${notificationId}/accept`, fetchConfig('POST', params), onSuccess, onErrors);
};

export const denyReadjustment = (notificationId, params, onSuccess, onErrors) => {
    appFetch(`/plans/notifications/${notificationId}/deny`, fetchConfig('POST', params), onSuccess, onErrors);
};

export const deleteTrainingSession = (sessionId, onSuccess, onErrors) => {
    appFetch(`/plans/training-sessions/${sessionId}`, fetchConfig('DELETE'), onSuccess, onErrors);
};

export const updateTrainingSession = (sessionId, params, onSuccess, onErrors) => {
    appFetch(`/plans/training-sessions/${sessionId}`, fetchConfig('PUT', params), onSuccess, onErrors);
};

File: src/main/java/es/udc/fi/dc/fd/model/services/PlanService.java
package es.udc.fi.dc.fd.model.services;

import es.udc.fi.dc.fd.model.entities.*;
import es.udc.fi.dc.fd.model.services.exceptions.IncorrectRoleException;
import es.udc.fi.dc.fd.model.services.exceptions.PermissionException;
import es.udc.fi.dc.fd.model.common.exceptions.DuplicateInstanceException;
import es.udc.fi.dc.fd.model.common.exceptions.InstanceNotFoundException;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;

public interface PlanService {
    DailyPlan getDailyPlan(Long userId, LocalDate date) throws InstanceNotFoundException;

    TrainingSession createTrainingSession(Long athleteId, Long coachId, LocalDate date, LocalTime startTime, TrainingSession.SportType sportType, String objective, String totalDistanceOrDuration, List<TrainingBlock> blocks) throws InstanceNotFoundException, IncorrectRoleException;

    NutritionPlan createNutritionPlan(Long athleteId, Long coachId, LocalDate date, Integer targetCalories, Integer proteinGrams, Integer carbsGrams, Integer fatGrams, Double hydrationLiters, String guidelines) throws InstanceNotFoundException, IncorrectRoleException, DuplicateInstanceException;

    RestPlan createRestPlan(Long athleteId, Long coachId, LocalDate date, Double targetSleepHours, String guidelines) throws InstanceNotFoundException, IncorrectRoleException, DuplicateInstanceException;

    TrainingBlock updateTrainingBlockDone(Long userId, Long blockId, Double done) throws InstanceNotFoundException, PermissionException;
    
    NutritionPlan updateNutritionPlanDone(Long userId, Long planId, Double done) throws InstanceNotFoundException, PermissionException;
    
    RestPlan updateRestPlanDone(Long userId, Long planId, Double done) throws InstanceNotFoundException, PermissionException;

    List<DailyPlan> getWeeklyPlan(Long userId, LocalDate startDate) throws InstanceNotFoundException;

    void rescheduleTrainingSession(Long userId, Long sessionId, LocalDate newDate, LocalTime newStartTime) throws InstanceNotFoundException, PermissionException;

    DailyPlan getAthleteDailyPlan(Long coachId, Long athleteId, LocalDate date) throws InstanceNotFoundException, PermissionException;

    List<DailyPlan> getAthleteWeeklyPlan(Long coachId, Long athleteId, LocalDate startDate) throws InstanceNotFoundException, PermissionException;

    List<Notification> getNotifications(Long coachId);
    
    void markNotificationAsRead(Long coachId, Long notificationId) throws InstanceNotFoundException, PermissionException;

    TrainingSession acceptReadjustment(Long coachId, Long userId, Long notificationId, Long sessionId, LocalDate newDate, LocalTime newStartTime, Boolean reschedule) throws InstanceNotFoundException, PermissionException;

    void denyReadjustment(Long coachId, Long userId, Long notificationId, Long sessionId, LocalDate newDate, LocalTime newStartTime) throws InstanceNotFoundException, PermissionException;

    Double calculateTSS(Long sessionId) throws InstanceNotFoundException, PermissionException;

    void deleteTrainingSession(Long coachId, Long sessionId) throws InstanceNotFoundException, PermissionException;

    TrainingSession updateTrainingSession(Long coachId, Long sessionId, LocalDate date, LocalTime startTime,
        TrainingSession.SportType sportType, String objective, String totalDistanceOrDuration, List<TrainingBlock> blocks) 
        throws InstanceNotFoundException, PermissionException;
    
    void deleteNutritionPlan(Long coachId, Long planId) throws InstanceNotFoundException, PermissionException;

    NutritionPlan updateNutritionPlan(Long coachId, Long planId, LocalDate planDate, Integer targetCalories, Integer proteinGrams, Integer carbsGrams,
        Integer fatGrams, Double hydrationLiters, String guidelines) throws InstanceNotFoundException, PermissionException;

    void deleteRestPlan(Long coachId, Long planId) throws InstanceNotFoundException, PermissionException;

    RestPlan updateRestPlan(Long coachId, Long planId, LocalDate planDate, Double targetSleepHours, String guidelines) throws InstanceNotFoundException, PermissionException;

    List<DailyPlan> getAthleteMonthlyPlan(Long coachId, Long athleteId, LocalDate startDate, LocalDate endDate) 
        throws InstanceNotFoundException, PermissionException;

    List<DailyPlan> getMonthlyPlan(Long userId, LocalDate startDate, LocalDate endDate) throws InstanceNotFoundException;

    TrainingReplanningProposal markSessionAsFailedAndReplan(Long userId, Long sessionId) throws InstanceNotFoundException, PermissionException;

    List<TrainingSession> acceptProposal(Long coachId, Long proposalId) throws InstanceNotFoundException, PermissionException;

    void denyProposal(Long coachId, Long proposalId) throws InstanceNotFoundException, PermissionException;

    TrainingReplanningProposal getPendingProposalBySessionId(Long userId, Long sessionId) throws InstanceNotFoundException, PermissionException;
}
